File: c_src/windows_authentication.c
/*

Copyright 2017 Neudesic, LLC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/

#include <Windows.h>
#include <erl_nif.h>

static const char* error_atom = "error";
static const char* invalid_token_handle_atom = "invalid_token_handle";
static const char* ok_atom = "ok";
static const char* win32_error_atom = "win32_error";

#define MAX_NAME 256

static HANDLE get_user_token(ErlNifEnv* env, ERL_NIF_TERM token_term) {
  HANDLE token_handle;

  return enif_get_ulong(env, token, (unsigned long*)&tokenHandle) ? token_handle : NULL;
}

static ERL_NIF_TERM make_win32_error_tuple(ErlNifEnv* env, DWORD error_code) {
  return enif_make_tuple2(
    env,
    enif_make_atom(env, error_atom),
    enif_make_tuple2(
      env,
      enif_make_atom(env, win32_error_atom),
      enif_make_atom(env, error_code)
    )
  );
}

static ERL_NIF_TERM make_invalid_token_handle_error_tuple(ErlNifEnv* env) {
  return enif_make_tuple2(
    env,
    enif_make_atom(env, error_atom),
    enif_make_atom(env, invalid_token_handle_atom)
  );
}

static ERL_NIF_TERM do_get_windows_username(ErlNifEnv* env, int argc, ERL_NIF_TERM argv[]) {
  HANDLE token_handle;
  DWORD token_user_length;
  PTOKEN_USER token_user;
  DWORD last_error;
  WCHAR username[MAX_NAME];
  DWORD username_length = MAX_NAME;
  WCHAR domain_name[MAX_NAME];
  DWORD domain_name_length = MAX_NAME;
  size_t converted_chars;
  char converted_username[MAX_NAME * 2];
  char converted_domain_name[MAX_NAME * 2];
  errno_t err;
  BOOL succeeded;
  SID_NAME_USE sid_name_use;

  token_handle = get_user_token(env, argv[0]);
  if (!token_handle) {
    return make_invalid_token_handle_error_tuple(env);
  }

  GetTokenInformation(
    token_handle,
    TokenUser,
    NULL,
    0,
    &token_user_length
  );
  last_error = GetLastError();
  if (ERROR_INSUFFICIENT_BUFFER != last_error) {
    return make_win32_error_tuple(env, last_error);
  }

  token_user = (PTOKEN_USER)malloc(token_user_length);
  succeeded = GetTokenInformation(
    token_handle,
    TokenUser,
    token_user,
    token_user_length,
    &token_user_length
  );
  if (!succeeded) {
    free_token(user);
    return make_win32_error_tuple(env, GetLastError());
  }

  succeeded = LookupAccountSidW(
    NULL,
    token_user->User.Sid,
    username,
    &username_length,
    domain_name,
    &domain_name_length,
    &sid_name_use
  );
  if (!succeeded) {
    free(token_user);
    return make_win32_error_tuple(env, GetLastError());
  }

  wcstombs_s(
    &converted_chars,
    converted_username,
    MAX_NAME * 2,
    username,
    username_length
  );
  wcstombs_s(
    &converted_chars,
    converted_domain_name,
    MAX_NAME * 2,
    domain_name,
    domain_name_length
  );

  free(token_user);
  return enif_make_tuple2(
    env,
    enif_make_atom(env, ok_atom),
    enif_make_tuple2(
      env,
      enif_make_string(env, converted_domain_name, ERL_NIF_LATIN1),
      enif_make_string(env, converted_username, ERL_NIF_LATIN1)
    )
  );
}

static ERL_NIF_TERM do_close_handle(ErlNifEnv *env, int argc, ERL_NIF_TERM argv[]) {
  HANDLE token_handle;

  token_handle = get_user_token(env, argv[0]);
  if (!token_handle) {
    return make_invalid_token_handle_error_tuple(env);
  }

  return CloseHandle(token_handle))
    ? enif_make_atom(env, ok_atom)
    : make_win32_error_tuple(env, GetLastError());
  }
}

static ErlNifFunc nif_functions[] = {
  { "do_close_handle", 1, do_close_handle },
  { "do_get_windows_username", 1, do_get_windows_username }
};

ERL_NIF_INIT(
  Elixir.Neudesic.WindowsAuthentication,
  nif_functions,
  NULL,
  NULL,
  NULL,
  NULL
);

File: lib/windows_authentication.ex
# Copyright 2017 Neudesic, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

defmodule Neudesic.WindowsAuthentication do
  @moduledoc """
  `Neudesic.WindowsAuthentication` implements a plug that will look for
  the `X-IIS-WindowsAuthToken` HTTP header in the request, and if found,
  will obtain the Windows domain user name for the authenticated user.
  This plug is intended to be used for enterprise intranet web applications
  that are built using the Phoenix framework and are deployed to a Windows
  server. The Phoenix web application will use IIS as a proxy and the
  [HttpPlatformHandler](https://www.iis.net/downloads/microsoft/httpplatformhandler)
  add-on to run the Phoenix web application and forward requests from IIS
  to the Phoenix application.

  `Neudesic.WindowsAuthentication` uses NIFs written in C to call Win32 APIs
  in order to lookup the domain name and user name for the user that was
  authenticated using integrated Windows authentication. The domain name
  and user name for the Windows user will be added to the
  `Conn.assigns` map.

  `Neudesic.WindowsAuthentication` is safe to use in all hosting environments.
  `Neudesic.WindowsAuthentication` will only process requests when the web
  application is being run on Windows. The NIFs will only be called if the
  `X-IIS-WindowsAuthToken` is present in the request. You can use the
  `Neudesic.WindowsAuthentication` plug when running from the command line,
  but Windows authentication and processing the request header will not
  occur unless the web application is running as a child process of IIS.
  """

  import Plug.Conn

  require Logger

  @windows_api Application.get_env(:neudesic_windows_auth_plug, :windows_api)

  @doc """
  Initializes the plug.
  """
  def init(%{:os_type => os_type}), do: os_type
  def init(_options), do: :os.type()

  @doc """
  Looks for the `X-IIS-WindowsAuthToken` HTTP header, and if present,
  stores the domain name and user name for the authenticated user in the
  `Conn.assigns` map using the `:windows_user` key.
  """
  def call(conn, os_type) do
    Logger.debug "Windows authentication plug called"
    do_windows_authentication(conn, os_type)
  end

  defp do_windows_authentication(conn, {:win32, _}) do
    do_process_windows_auth_token_header(conn, get_req_header(conn, "x-iis-windowsauthtoken"))
  end

  defp do_windows_authentication(conn, _) do
    Logger.debug "The Windows authentication plug is not running on Windows; skipping"
    conn
  end

  defp do_process_windows_auth_token_header(conn, [token_string]) do
    Logger.debug "Found the X-IIS-WindowsAuthToken HTTP header"
    Logger.debug fn -> "Windows authentication token = #{token_string}" end

    token_handle = String.to_integer(token_string, 16)
    
    Logger.debug "Getting the Windows user name"
    {:ok, windows_user} = @windows_api.get_windows_username(token_handle)
    Logger.debug fn ->
      {domain, username} = windows_user
      "Windows user = #{domain}\\#{username}"
    end
    
    Logger.debug "Closing the Windows authentication token handle"
    @windows_api.close_handle(token_handle)
    
    Logger.debug "Saving the Windows user name to Conn.assigns"
    assign(conn, :windows_user, windows_user)
  end

  defp do_process_windows_auth_token_header(conn, []) do
    Logger.debug "The X-IIS-WindowsAuthToken HTTP header was not present on the request"
    conn
  end
end